_posts/2023-06-20-introducing-iam-for-openfaas.md

```diff
@@ -0,0 +1,479 @@
+- If you are deploying the OpenFaaS dashboard add the redirect URI for your dashboard, e.g `https://dashboard.openfaas.example.com/auth/callback`
+
+
+### Register a provider with OpenFaaS
+
+There must be at least one registered OIDC provider for human users to authenticate but additional issuers can be defined for e.g. [Web Identity Federation](https://docs.openfaas.com/openfaas-pro/iam/github-actions-federation/)
+
+A provider can be registered with OpenFaaS by creating a JwtIssuer object in the `openfaas` namespace. JwtIssuer objects are used to define trusted issuers for OpenFaaS IAM.
+
+Example issuer for a Keycloak provider:
+
+```yaml
+---
+apiVersion: openfaas.com/v1
+kind: JwtIssuer
+metadata:
+  name: keycloak.example.com
+  namespace: openfaas
+spec:
+  iss: https://keycloak.example.com/realms/openfaas
+  aud:
+    - openfaas
+  tokenExpiry: 12h
+```
+
+In the spec the `iss` field needs to be set to the url of your provider, eg. `https://accounts.google.com` for google or `https://example.eu.auth0.com/ for Auth0.
+
+The `aud` field contains a list of client apps this JwtIssuer is valid for. The value should be the client_id you received while registering an app with your IdP.
+
+In addition to the JwtIssuer there are two other object that need to be defined before we can try and access the gateway API:
+
+ - Policy - defines a set of permissions and objects on which they can be performed
+ - Role - defines a set of policies that can be matched to a particular user or identity
+
+### Create roles and policies
+
+Once a JWTIssuer has been registered you can start creating Roles and Policies. Policies are used to describe permissions on resources. A Role must be created to map users within the Issuer to a set of Policies based on conditions.
+
+Create a Role:
+```yaml
+apiVersion: openfaas.com/v1
+kind: Role
+metadata:
+  name: dev-staff-deployers
+  namespace: openfaas
+spec:
+  policy:
+  - dev-rw
+  principal:
+    jwt:sub:
+     - aa544816-e4e9-4ea0-b4cf-dd70db159d2e
+  condition:
+    StringEqual:
+      jwt:iss: [ "https://keycloak.example.com/realms/openfaas" ]
+```
+
+The `policy` field contains a set of Policies to apply for this Role.
+
+The `condition` field can be used to limit permissions by matching fields in the jwt access token.
+
+Every condition must return true for the Role to be considered as a match.
+
+Valid conditions include: `StringEqual` or `StringLike`
+
+A user's email could also be fuzzy matched with a condition, for example:
+
+```yaml
+condition:
+    StringLike:
+      jwt:email: ["*@example.com"]
+```
+
+The principal filed is optional, however if it is given, both the principal and the condition must match. If the principal contains multiple items only one must match the token for it to be considered a match.
+
+In this example an access token will match the Role for 1 staff member with sub `aa544816-e4e9-4ea0-b4cf-dd70db159d2e` only if it was issued by `https://keycloak.example.com/realms/openfaas`.
```

Comment:
In which example?

The policy or the role?

How about: "This Role will only match for ...  "

Reply:
staff member => user